alibi/explainers/ale.py

```diff
@@ -0,0 +1,479 @@
+                                      constant_value=constant_value,
+                                      feature_values=feature_values,
+                                      feature_deciles=feature_deciles)
+
+    def build_explanation(self,
+                          ale_values: List[np.ndarray],
+                          ale0: List[np.ndarray],
+                          constant_value: float,
+                          feature_values: List[np.ndarray],
+                          feature_deciles: List[np.ndarray]) -> Explanation:
+        """
+        Helper method to build the Explanation object.
+        """
+        # TODO decide on the format for these lists of arrays
+        # Currently each list element relates to a feature and each column relates to an output dimension,
+        # this is different from e.g. SHAP but arguably more convenient for ALE.
+
+        data = copy.deepcopy(DEFAULT_DATA_ALE)
+        data.update(ale_values=ale_values,
+                    ale0=ale0,
+                    constant_value=constant_value,
+                    feature_values=feature_values,
+                    feature_names=self.feature_names,
+                    target_names=self.target_names,
+                    feature_deciles=feature_deciles)
+
+        return Explanation(meta=copy.deepcopy(self.meta), data=data)
+
+
+def get_quantiles(values: np.ndarray, num_points: int = 11, interpolation='linear') -> np.ndarray:
+    """
+    Calculate quantiles of values in an array.
+
+    Parameters
+    ----------
+    values
+        Array of values.
+    num_points
+        Number of quantiles to calculate.
+
+    Returns
+    -------
+    Array of quantiles of the input values.
+
+    """
+    percentiles = np.linspace(0, 100, num=num_points)
+    quantiles = np.percentile(values, percentiles, axis=0, interpolation=interpolation)
+    return quantiles
+
+
+def bisect_fun(fun: Callable, target: float, lo: int, hi: int) -> int:
+    """
+    Bisection algorithm for function evaluation with integer support.
+
+    Assumes the function is non-decreasing on the interval [lo, hi].
+    Return an integer value v such that for all x<v, fun(x)<target and for all x>=v fun(x)>=target.
+    This is equivalent to the library function `bisect.bisect_left` but for functions defined on integers.
+
+    Parameters
+    ----------
+    fun
+        A function defined on integers in the range [lo, hi] and returning floats.
+    target
+        Target value to be searched for.
+    lo
+        Lower bound of the domain.
+    hi
+        Upper bound of the domain.
+
+    Returns
+    -------
+    Integer index.
+
+    """
+    while lo < hi:
+        mid = (lo + hi) // 2
+        if fun(mid) < target:
+            lo = mid + 1
+        else:
+            hi = mid
+    return lo
+
+
+def adaptive_grid(values: np.ndarray, min_bin_points: int = 1) -> Tuple[np.ndarray, int]:
+    """
+    Find the optimal number of points to subdivide the feature range into
```

Comment:
Replaced with `    Find the optimal number of quantiles for the range of values so that each resulting bin
    contains at least `min_bin_points`. Uses bisection.`